### PR TITLE
Rename `hash` module to `merkle_root`

### PR DESCRIPTION
**Pull request by tcharding**
Done as part of flattening `util`.

The `util::hash` module only provides two functions, both to calculate the merkle root of a list of hashes.

1. Rename `util::hash` -> `crate::merkle_root`
2. Change function names to `calculate[_inline]` so usage becomes `merkle_root::calculate`

Done as two separate patches so we can bikeshed the names, can squash if needed.

### CODE REVIEW

**Comment by apoelstra** [comment]
concept ACK. If #418 gets in before this one, those new functions should also live in the new `merkle_root` module.

**Review by apoelstra — APPROVED** [review]
ACK bc845d08dc4917a1a2d5b54901507d7bbb506651

**Comment by tcharding** [comment]
> concept ACK. If #418 gets in before this one, those new functions should also live in the new `merkle_root` module.

#418 was merged two years ago ;) Did you mean #1321?

**Comment by apoelstra** [comment]
Oops, yes. I blame github's # autocomplete :).

**Comment by Kixunil** [comment]
How about module being named `merkle_tree` and functions being `calculate_root`?

**Comment by tcharding** [comment]
I like it!

**Comment by tcharding** [comment]
Renamed as suggested.

**Comment by apoelstra** [comment]
CI failure is actually in master, sorry

**Comment by tcharding** [comment]
No sweat, I'll fix it :)

**Comment by apoelstra** [comment]
Sure -- though I have a fix that I'll push as soon as my local tests pass. (If you beat me to it, fine, I'll just test and ACK yours -- it's only a couple lines :))

**Comment by tcharding** [comment]
https://github.com/rust-bitcoin/rust-bitcoin/pull/1346 whichever goes first :)

**Comment by tcharding** [comment]
Rebased on master, no other changes.

**Review by apoelstra — DISMISSED** [review]
ACK a7e74fc7355aab932fc68c4896593c982e5a320b

File: bitcoin/src/merkle_tree.rs [hunk 1]
```diff
@@ -19,13 +26,13 @@ use crate::consensus::encode::Encodable;
...
 ///
-/// In most cases, you'll want to use [bitcoin_merkle_root] instead.
+/// In most cases, you'll want to use [`calculate_root`] instead.
 ///
 /// # Returns
 /// - `None` if `hashes` is empty. The merkle root of an empty tree of hashes is undefined.
 /// - `Some(hash)` if `hashes` contains one element. A single hash is by definition the merkle root.
 /// - `Some(merkle_root)` if length of `hashes` is greater than one.
-pub fn bitcoin_merkle_root_inline<T>(hashes: &mut [T]) -> Option<T>
+pub fn calculate_root_inline<T>(hashes: &mut [T]) -> Option<T>
```
Comment (Kixunil) [inline bitcoin/src/merkle_tree.rs:35]: Now that I see this I realize it'd be nice to explicitly state that the hashes are overwritten in the process.  And perhaps state that the state we leave it in is undefined and must not be relied on.

File: bitcoin/src/util/mod.rs [hunk 2]
```diff
@@ -97,6 +96,34 @@ pub use crate::bip32;
 #[deprecated(since = "0.30.0", note = "Please use crate::bip158")]
 pub use crate::bip158;
 
+/// The `hash` module was renamed to `../merkle_tree`.
```
Comment (Kixunil) [inline bitcoin/src/util/mod.rs:99]: Although GH says "renamed" and we think of it as renaming, a more sensible representation is "removed and the merkle tree functions were moved to `merkle_tree`"

File: bitcoin/src/util/mod.rs [hunk 3]
```diff
@@ -97,6 +96,34 @@ pub use crate::bip32;
...
+    pub fn bitcoin_merkle_root<T, I>(hashes: I) -> Option<T>
+    where
+        T: Hash + Encodable,
+    <T as Hash>::Engine: io::Write,
+        I: Iterator<Item=T>,
+    {
+        merkle_tree::calculate_root(hashes)
+    }
+}
+
```
Comment (Kixunil) [inline bitcoin/src/util/mod.rs:126]: Very good choice here finding them could be hard otherwise.

**Review by Kixunil — DISMISSED** [review]
ACK a7e74fc7355aab932fc68c4896593c982e5a320b

File: bitcoin/src/merkle_tree.rs (hunk 1 above)
Comment (apoelstra) [reply bitcoin/src/merkle_tree.rs:35]: Agreed. Though FWIW taking a mutable slice implies that we'll overwrite stuff.

File: bitcoin/src/util/mod.rs (hunk 2 above)
Comment (tcharding) [reply bitcoin/src/util/mod.rs:99]: Noted, all thees "deprecated" messages should go away in a version or two so unless I re-spin for some other reason I'll leave this as is.

File: bitcoin/src/util/mod.rs (hunk 3 above)
Comment (tcharding) [reply bitcoin/src/util/mod.rs:126]: Cheers!

File: bitcoin/src/util/mod.rs (hunk 2 above)
Comment (tcharding) [reply bitcoin/src/util/mod.rs:99]: Changed to: `/// Functions from the `hash` module were renamed and moved to `../merkle_tree`.`

File: bitcoin/src/merkle_tree.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/merkle_tree.rs:35]: Added some extra docs as an additional patch.

**Comment by tcharding** [comment]
Changes in force push:
- rebased fixing trivial merge conflict
- added additional patch with improved rustdocs

**Review by Kixunil — APPROVED** [review]
ACK 29df410ea330c9a0feae8416c826d3711a3502bf

**Review by apoelstra — APPROVED** [review]
ACK 29df410ea330c9a0feae8416c826d3711a3502bf

**Pull request merged**